### PR TITLE
Add keypad usecase

### PR DESCRIPTION
**Pull request by cedelavergne-ledger**
## Description

Add functions and definitions to support a standard default keypad usecase.

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Tests
- [ ] Documentation
- [ ] Other (for changes that might not fit in any category)

### CODE REVIEW

File: lib_nbgl/src/nbgl_use_case.c [hunk 1]
```diff
@@ -1526,5 +1547,197 @@ void nbgl_useCaseSpinner(const char *text)
...
+#ifdef NBGL_KEYPAD
+/**
+ * @brief internal function to refresh the keypad and automatically show / hide:
+ *        - backspace key if no digits are present
+ *        - validation key if the min digit is reached
+ *        - keypad if the max number of digit is reached
+ *
+ * @param add indicate if a digit has been added or removed
+ */
+static void nbgl_refreshKeyPad(bool add)
```
Comment (nroggeman-ledger) [inline lib_nbgl/src/nbgl_use_case.c:1560]: Please follow the current syntax and do not use the "refresh" word, already used for "screen refresh"... nbgl_useCaseUpdateKeypad()

File: lib_nbgl/src/nbgl_use_case.c [hunk 2]
```diff
@@ -1526,5 +1547,197 @@ void nbgl_useCaseSpinner(const char *text)
...
+    nbgl_refreshSpecialWithPostRefresh(mode, POST_REFRESH_FORCE_POWER_ON);
+}
+
+/**
+ * @brief keypad callback when a key is touched
+ *        - backspace key
+ *
+ * @param touchedKey the key touched
+ */
+static void nbgl_keypadCallback(char touchedKey)
```
Comment (nroggeman-ledger) [inline lib_nbgl/src/nbgl_use_case.c:1610]: to be complient with all NBGL files, put static functions before all global functions

File: lib_nbgl/src/nbgl_use_case.c [hunk 3]
```diff
@@ -1526,5 +1547,197 @@ void nbgl_useCaseSpinner(const char *text)
...
+    keypadContext.pinMaxDigits = maxDigits;
+
+    nbgl_layoutDraw(keypadContext.layoutCtx);
+    nbgl_refreshSpecialWithPostRefresh(FULL_COLOR_CLEAN_REFRESH, POST_REFRESH_FORCE_POWER_ON);
+}
+
+/**
+ * @brief keypad layout release
+ */
+void nbgl_useCaseKeypadRelease(void)
```
Comment (nroggeman-ledger) [inline lib_nbgl/src/nbgl_use_case.c:1733]: Why a release for Keypad ? Use Case layer doesn't require it....

File: lib_nbgl/src/nbgl_use_case.c (hunk 1 above)
Comment (nroggeman-ledger) [reply lib_nbgl/src/nbgl_use_case.c:1560]: Sorry, if it's a static function, it can be have any name

Comment (cedelavergne-ledger) [reply lib_nbgl/src/nbgl_use_case.c]: This function is `static`. I have renamed it `updateKeyPad`.

File: lib_nbgl/src/nbgl_use_case.c (hunk 2 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/src/nbgl_use_case.c]: Done

File: lib_nbgl/src/nbgl_use_case.c (hunk 3 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/src/nbgl_use_case.c]: Ok, removed

**Review by nroggeman-ledger — APPROVED** [review]

File: lib_nbgl/doc/nbgl_use_case.dox [hunk 4]
```diff
@@ -866,6 +868,56 @@ An extra button under the raw address enables to open a modal page to see the ad
...
+- a boolean to request a shuffled keypad
+- a tune value
+- callbacks for navigation and pin validation
+
+@note The \em backspace and \em validate buttons will be shown or hidden automatically.
+
+Here is the code to display something similar to example picture:
+
+@code
+static void validate_pin(uint8_t *pinentry, uint8_t length) {
```
Comment (xchapron-ledger) [inline lib_nbgl/doc/nbgl_use_case.dox:896]:
```suggestion
static void validate_pin(const uint8_t *pinentry, uint8_t length) {
```?

File: lib_nbgl/doc/nbgl_use_case.dox [hunk 5]
```diff
@@ -866,6 +868,56 @@ An extra button under the raw address enables to open a modal page to see the ad
...
+    UNUSED(index);
+    // Callback for the key navigation (back key mainly)
+    if (token == TOKEN_PIN_ENTRY_BACK) {
+        ui_init();
+    }
+}
+
+void ui_menu_pinentry_display(unsigned int value) {
+    // Draw the keypad
+    nbgl_useCaseKeypad("Enter the User",
```
Comment (xchapron-ledger) [inline lib_nbgl/doc/nbgl_use_case.dox:910]:
```suggestion
    nbgl_useCaseKeypad("Enter User PIN",
```
? to match the png?

File: lib_nbgl/include/nbgl_use_case.h [hunk 6]
```diff
@@ -107,6 +107,13 @@ typedef void (*nbgl_choiceCallback_t)(bool confirm);
  */
 typedef void (*nbgl_actionCallback_t)(uint8_t page);
 
+/**
+ * @brief prototype of pin validation callback function
+ * @param content pin value
+ * @param length pin length
+ */
+typedef void (*nbgl_pinValidCallback_t)(uint8_t *content, uint8_t page);
```
Comment (xchapron-ledger) [inline lib_nbgl/include/nbgl_use_case.h:115]:
```suggestion
typedef void (*nbgl_pinValidCallback_t)(const uint8_t *content, uint8_t page);
```
?

File: lib_nbgl/src/nbgl_use_case.c (hunk 3 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/src/nbgl_use_case.c:1723]: Indeed, it is needed, else, the screen is kept in background (not released)

File: lib_nbgl/doc/nbgl_use_case.dox (hunk 4 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/doc/nbgl_use_case.dox:897]: Fixed

File: lib_nbgl/include/nbgl_use_case.h (hunk 6 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/include/nbgl_use_case.h:115]: Fixed

**Review by nroggeman-ledger — APPROVED** [review]

File: lib_nbgl/src/nbgl_use_case.c (hunk 3 above)
Comment (cedelavergne-ledger) [reply lib_nbgl/src/nbgl_use_case.c]: Ok, removing `modal`, and now the release is no more needed.
Fixed!

**Comment by aido** [comment]
Hi @cedelavergne-ledger.

This default keypad use case seems to be suitable only for PINs i.e. hidden digits. What if a user needs to use a keypad to input a number (not PIN) within a specific range?
Can `nbgl_useCaseKeypad()` be made more generic so that digits are NOT hidden? Or a possible split of `nbgl_useCaseKeypad()`  into `nbgl_useCaseKeypadDigits()` and `nbgl_useCaseKeypadPIN()` functions?

**Comment by cedelavergne-ledger** [comment]
> Hi @cedelavergne-ledger,
> 
> This default keypad use case seems to be suitable only for PINs i.e. hidden digits. What if a user needs to use a keypad to input a number (not PIN) within a specific range? Can `nbgl_useCaseKeypad()` be made more generic so that digits are optionally NOT hidden? Or a possible split of `nbgl_useCaseKeypad()` into `nbgl_useCaseKeypadDigits()` and `nbgl_useCaseKeypadPIN()` functions?

Hi @aido 

This is a good question. Effectively, it has not been written for generic numbers. It could optimized that way, or using an additional parameter in the function.
Please note you can also call the different functions to directly draw your keypad.

**Pull request merged**